src/Framework/PaymentGateways/Routes/GatewayRoute.php

```diff
@@ -2,74 +2,94 @@
 
 /**
  * @since 2.18.0
  */
 class GatewayRoute
 {
+    use ResponseHelpers;
+
     /**
-     * @since 2.18.0
+     * This is our entry point into the Gateway Routing system.
      *
      * @return void
      * @throws PaymentGatewayException
+     * @since 2.18.0
+     * @unreleased - validate secureRouteMethods
+     *
      */
     public function __invoke()
     {
         if ($this->isValidListener()) {
             /** @var PaymentGatewayRegister $paymentGatewaysRegister */
             $paymentGatewaysRegister = give(PaymentGatewayRegister::class);
+
+            // get all registered gateways
             $paymentGateways = $paymentGatewaysRegister->getPaymentGateways();
+
+            // get all registered gateway ids
             $gatewayIds = array_keys($paymentGateways);
 
+            // make sure required params are valid
             if (!$this->isValidRequest($gatewayIds)) {
                 throw new PaymentGatewayException('This route is not valid.');
             }
 
-            $data = GatewayRouteData::fromRequest($_GET);
+            // create DTO from GET request
+            $data = GatewayRouteData::fromRequest(give_clean($_GET));
 
-            /** @var PaymentGateway $gateway */
+            /**
+             * Get the PaymentGateway instance
+             *
+             * @var PaymentGateway $gateway
+             */
             $gateway = give($paymentGateways[$data->gatewayId]);
 
-            $allowedGatewayMethods = $gateway->routeMethods;
-
-            if (is_a($gateway, OffsiteGatewayInterface::class)) {
-                $allowedGatewayMethods = array_merge(
-                    $allowedGatewayMethods,
-                    OffsiteGatewayInterface::defaultRouteMethods
-                );
-            }
+            // get all public and secure gateway methods
+            $allGatewayMethods = array_merge($gateway->routeMethods, $gateway->secureRouteMethods);
 
+            // Make sure the method being called is defined in the gateway.
             if (
-                !in_array($data->gatewayMethod, $allowedGatewayMethods, true) ||
-                !method_exists($gateway, $data->gatewayMethod)
+                !method_exists($gateway, $data->gatewayMethod) ||
+                !in_array($data->gatewayMethod, $allGatewayMethods, true)
             ) {
                 throw new PaymentGatewayException('The gateway method does not exist.');
             }
 
-            $gateway->handleGatewayRouteMethod($data->donationId, $data->gatewayMethod);
+            // If method is in secureRouteMethods then, validate signature
+            if (in_array($data->gatewayMethod, $gateway->secureRouteMethods, true)) {
+                $this->validateSignature($data->routeSignature, $data);
+            }
+
+            // Navigate to our payment gateway api to handle calling the gateway's method
+            $this->handleGatewayRouteMethod($gateway, $data->gatewayMethod, $data->queryParams);
```

Comment:
@jonwaldstein this will last point of the script after that we do not want anything to run. Can will add `exit` after this?

Reply:
@ravinderk Is that necessary? I would think the try...catch inside the function compensates for finishing the request.

Reply:
@jonwaldstein Yes, I think so.
Process a dummy gateway route then you will see that home page HTML will load if the payment gateway does not use exit in their gateway method.
With my suggestion, we can prevent code execution after this point.

Reply:
@ravinderk would you mind sharing a screenshot? I'm not able to reproduce

Reply:
![image](https://user-images.githubusercontent.com/1784821/152017436-2e43f8af-1d84-4391-9495-1c2b6ef94789.png)

Reply:
@ravinderk exit added 3c53f012